Groups: complete homomorphisms

valglad opened the pull request:
This PR adds the remaining homomorphism cases (with `PermutationGroup` as the domain or `FpGroup`/`FreeGroup` as the codomain). Checking if a homomorphism from a permutation group is well-defined is not yet implemented but these homomorphisms can be created by setting the keyword argument `check` in the `homomorphism` function to `False`.

Other changes:
*  `random_element` method for `FpGroup`s was renamed as `random` for consistency with permutation groups
*  `generator_product` gets a keyword argument `original` so that the returned list contains only the original generators (the default is strong generators)
*  fixed a minor bug in `coset_unrank`: at the moment it doesn't work unless `schreier_sims` is run before calling it

valglad committed:
add other homomorphism cases

valglad committed:
merge master

jksuom commented on sympy/combinatorics/homomorphisms.py:
```diff
@@ -36,35 +36,57 @@ def _invs(self):
-            for s in image._strong_gens_slp[g]:
+            if isinstance(self.codomain, PermutationGroup):
+                parts = image._strong_gens_slp[g][:]
+                parts = [parts[len(parts)-i-1] for i in range(len(parts))]
```
A list can be reversed by `[::-1]`:
```
>>> l = [1, 2, 3, 4]
>>> l[::-1]
[4, 3, 2, 1]
```

jksuom commented on sympy/combinatorics/homomorphisms.py:
```diff
@@ -36,35 +36,57 @@ def _invs(self):
+        else:
+            gens = g
+        for i in range(len(gens)):
+            s = gens[i]
```
Could this be `for s in gens:`?

valglad replied:
I think so. I don't remember why I used `i` here.

valglad replied:
I remembered why. Because `for s in gens:` would be equivalent to `for s in gens.array_form:` when `g` is a `FreeGroupElement`.  But when you call `gens` by index, the generator (or inverse) at position `i` is returned. It's actually a bit inconsistent. Maybe something needs to be done about it.

jksuom replied:
This seems to be another place where a comment could be helpful.

jksuom commented on sympy/combinatorics/homomorphisms.py:
```diff
@@ -36,35 +36,57 @@ def _invs(self):
+            if s in self._inverses:
+                w = w*self._inverses[s]
+            else:
+                w = w*self._inverses[s**-1]**-1
```
Is `s**-1` in `self._inverses` if `s` isn't?

valglad replied:
Yes, `_inverses` contains all strong generators or (or just regular generators, if it's not a permutation group) and `s` is one of them or the inverse.

jksuom commented on sympy/combinatorics/homomorphisms.py:
```diff
@@ -103,28 +131,39 @@ def image(self):
+                gens = self.domain.generator_product(elem, original=True)
+                for g in gens:
+                    if g in self.images:
+                        value = images[g]*value
```
How are the items of `generator_product` ordered? A homomorphism should preserve the permutation group product order (which is the inverse of the standard composition of mappings).

valglad replied:
If `generator_product(elem)` returns `[s_1 ... s_n]`, `g == s_n*...*s_1`, and `value = images[g]*value` seems to be consistent with it.

jksuom commented on sympy/combinatorics/homomorphisms.py:
```diff
@@ -176,10 +215,16 @@ def homomorphism(domain, codomain, gens, images=[]):
-        raise NotImplementedError("Homomorphisms from permutation groups are not currently implemented")
-    elif not isinstance(domain, (FpGroup, FreeGroup)):
+    if check and isinstance(domain, PermutationGroup):
+        raise NotImplementedError("Checking if the homomorphism is well-defined\
```
It is better to build the string by concatenation. Otherwise all the whitespace in the beginning of lines will be included.
```
    NotImplementedError("Checking if the homomorphism is well-defined "  # note the space at end
               "<next line with ending space>"
               "<last line>")

jksuom commented on sympy/combinatorics/homomorphisms.py:
```diff
@@ -199,29 +244,42 @@ def homomorphism(domain, codomain, gens, images=[]):
             r_arr = r.array_form
             i = 0
+            j = 0
             while i < len(r):
```
I think a comment should be added to explain why both forms of `r` are needed. (Maybe the  definition  of array form should be changed to consist of pairs `(gen, exp)`.)

valglad replied:
That would be more convenient for loops but I'm not sure if it is possible. At the moment, a group element is basically defined as an array of tuples like this (with appropriate operations for combining them). If `gen` is made to be a `FreeGroupElement` itself, then group elements might end up being defined in terms of themselves and that sounds circular so could cause problems. Or maybe it'll be fine. I'll look into it. Even if it isn't possible, it could be convenient to have another attribute like that.

jksuom commented on sympy/combinatorics/homomorphisms.py:
```diff
@@ -199,29 +244,42 @@ def homomorphism(domain, codomain, gens, images=[]):
     return GroupHomomorphism(domain, codomain, images)
 
-def _check_homomorphism(domain, images, identity):
+def _check_homomorphism(domain, codomain, images):
```
Is this going to be used only for permutation group domains?

valglad replied:
For any domains, unless `check==False`. It depends on knowing the relators of the group so it would have to find an `FpGroup` presentation of a permutation group first.

jksuom commented on sympy/combinatorics/homomorphisms.py:
```diff
@@ -199,29 +244,42 @@ def homomorphism(domain, codomain, gens, images=[]):
+                # only try to make the rewriting system
+                # confluent when it can't determine the
+                # truth of equality otherwise
+                codomain.make_confluent()
```
There should probably be some way of checking to see if this was successful.

valglad replied:
Yes, maybe `make_confluent` could be made to return a boolean.

valglad committed:
temp

jksuom commented on sympy/combinatorics/tests/test_homomorphisms.py:
```diff
@@ -21,12 +22,30 @@ def test_homomorphism():
+    # FreeGroup -> FpGroup
+    G = FpGroup(F, [a*b*a**-1*b**-1])
+    T = homomorphism(F, G, F.generators, G.generators)
+    assert T.invert(a**-1*b**-1*a**2) == a**-1*b**-1*a**2
```
`G.reduce` doesn't seem to work as expected. The reduced form of `a**-1*b**-1*a**2` should be `a*b**-1`. Apparently the rules should be further simplified. Currently they are

    {b*a: a*b, b*a**-1*b**-1: a**-1, a*b*a**-1: b, b**-1*a**-1: a**-1*b**-1}

The second and third one should probably be `b*a**-1: a**-1*b` and `b**-1*a: a*b**-1`.

valglad replied:
It should be `a*b**-1` if `G._rewriting_system` is confluent. But `make_confluent` can take a while so by default it isn't. As for  `b*a**-1: a**-1*b `and `b**-1*a: a*b**-1`, they should have been added to the dictionary when the rules were initiated. I wonder why they weren't. Will work on it.

jksuom replied:
It seems that one rule is still missing:
```
(Pdb) p rules
{b*a: a*b, b*a**-1: a**-1*b, b**-1*a**-1: a**-1*b**-1}
(Pdb) p self.is_confluent
True
```

valglad committed:
rewriting system changes and comments

add_rule method changed to directly add all rules that can be obtained from
a given pair of words (it wasn't a major issue as the other rules
would eventually be found by make_confluent but it is probably more
efficient this way). Overlaps in make_confluent should be found with
the inverses of the rules as well as their regular versions. The change
in the reduce method was made because the previous "if" condition could
lead to problems (e.g. with a hypothetical rule b -> a). symbols_index
method for free groups was only used in one place in the code and didn't
do what intented (because negative of 0 is 0) so it is removed and the
__le__ method of FreeGroupElements is fixed. The other changes
are to do with shortening the code or adding comments.

valglad committed:
free group changes

valglad committed:
correct a typo

valglad committed:
fix add_rule

another case to check for was added to add_rule method. _add_rule
(directly adding the rule with no processing) has been added for
convenience.

valglad committed:
eliminate_word should preserve inverse

jksuom commented on sympy/combinatorics/rewritingsystem.py:
```diff
@@ -80,40 +91,33 @@ def add_rule(self, w1, w2, check=False):
+                new_keys.add(s1)
+                if not check:
+                    self._add_rule(s1, s2)
+            if s2**-1 > s1**-1 and s2**-1 not in self.rules:
```
What should be done if `s1**-1 > s2**-1`? Maybe there should be a more systematic way of adding these 'inverse' rules.

valglad replied:
In that case it is accounted for by s1 -> s2. The reduction process will substitute for both s1 and s1**-1 except when s2**-1 > s1**-1  so I think that is the only case that needs special treatment.

valglad committed:
another rule case and rules cache

jksuom commented on sympy/combinatorics/rewritingsystem.py:
```diff
@@ -32,6 +33,8 @@ def __init__(self, group):
 
         # dictionary of reductions
         self.rules = {}
+        self.rules_cache = deque([])
```
It is possible to set `maxlen` of a deque. Then its size is automatically bounded.

jksuom commented on sympy/combinatorics/rewritingsystem.py:
```diff
@@ -135,11 +163,8 @@ def _remove_redundancies(self, changes=False):
-                    added.update(new)
-            else:
-                self.rules[v] = w
+            new = self.add_rule(v, w)
```
What is the purpose of this line? The rule `v -> w` is in `self.rules` and this call may reintroduce other redundant rules that have already been deleted in the loop.

valglad replied:
I should probably change that and use `_add rule` when `v != r`.

valglad replied:
Actually, using `add_rule` to make deductions from the reduced rules in place seems to be faster. I'm going to allow it during `make_confluent` (it's only used to tidy up a bit, so it's not important that the system is completely reduced) with a keyword argument but otherwise it shouldn't do that.

valglad replied:
Nevermind, that was probably due to randomness.

valglad committed:
fix remove_redundancies
